test/src/test/java/org/corfudb/runtime/object/transactions/OptimisticTransactionContextTest.java

```diff
@@ -842,4 +867,35 @@ public void txnSpansEpochsCanCommitIfPrimarySequencerSpansConsecutiveEpochs() th
         // When t2 wants to commit, the sequencer does not have consecutive epochs, so it should abort.
         t2(this::TXEnd).assertThrows().isInstanceOf(TransactionAbortedException.class);
     }
+
+    /**
+     * Test that the transaction can commit if the first attempt got overwritten and retried.
+     * Basically on overwrite and retry we need to adjust the transaction snapshot otherwise
+     * the transaction will always conflict with itself and get TransactionAbortedException.
+     */
+    @Test
+    public void txnCanCommitIfOverwrittenAndRetry() throws Exception {
+        // Get a new runtime.
+        CorfuRuntime rt = getNewRuntime(getRuntime().getParameters())
+                .parseConfigurationString(getDefaultConfigurationString());
+        rt.connect();
+
+        OptimisticTXBegin();
+
+        put("k1", "v1");
+        put("k2", "v2");
+
+        // Add a client rule that does a hole fill to create an OverwriteException
+        // after the token is issued with no conflict and before writing to log unit.
+        AtomicInteger counter = new AtomicInteger(0);
```

Comment:
A cleaner way is to just fill a hole directly without going to the sequencer such that the next token handed out will cause the writer to fail and retry. Its much cleaner than this type of synchronization

Reply:
Done.